lib/src/main/java/graphql/nadel/engine/NadelIncrementalResultSupport.kt

```diff
@@ -0,0 +1,181 @@
+package graphql.nadel.engine
+
+import graphql.incremental.DelayedIncrementalPartialResult
+import graphql.nadel.engine.NadelIncrementalResultSupport.OutstandingJobCounter.OutstandingJobHandle
+import graphql.nadel.engine.util.copy
+import graphql.nadel.util.getLogger
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.launch
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * todo: we do not handle the case where defer jobs finish before [onInitialResultComplete]
+ */
+class NadelIncrementalResultSupport internal constructor(
+    private val delayedResultsChannel: Channel<DelayedIncrementalPartialResult> = Channel(
+        capacity = 100,
+        onBufferOverflow = BufferOverflow.DROP_LATEST,
+        onUndeliveredElement = {
+            log.error("Dropping incremental result because of buffer overflow")
+        },
+    ),
+) {
+    companion object {
+        private val log = getLogger<NadelIncrementalResultSupport>()
+    }
+
+    /**
+     * The root [Job] to run the defer and stream work etc on.
+     */
+    private val coroutineJob = SupervisorJob()
+    private val coroutineScope = CoroutineScope(coroutineJob + Dispatchers.Default)
+
+    /**
+     * A single [Flow] that can only be collected from once.
+     */
+    private val resultFlow by lazy(delayedResultsChannel::consumeAsFlow)
+
+    init {
+        coroutineJob.invokeOnCompletion {
+            require(outstandingJobCounter.isEmpty())
+            delayedResultsChannel.close()
+        }
+    }
+
+    private val outstandingJobCounter = OutstandingJobCounter()
+
+    fun defer(task: suspend CoroutineScope.() -> DelayedIncrementalPartialResult): Job {
+        return launch { outstandingJobHandle ->
+            val hasNext: Boolean
+            val result = try {
+                task()
+            } finally {
+                hasNext = outstandingJobHandle.decrementAndGetJobCount() > 0
+            }
+
+            delayedResultsChannel.send(
+                // Copy of result but with the correct hasNext according to the info we know
+                result.copy(hasNext = hasNext)
+            )
+        }
+    }
+
+    /**
+     * Assumes that the [Flow] will correctly set [DelayedIncrementalPartialResult.hasNext] correctly.
```

Comment:
@sbarker2 @Juliano-Prado This is the method that the "non-hydrated" defer code path will use. You don't have to worry about what the value of `hasNext` is, as this should have been set correctly by the underlying service, though.